### ISSUE TITLE
feat: Parameter groups — exclude BatchNorm from weight decay

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#209]
## Summary

AdamW applies weight decay uniformly to **all** parameters, including BatchNorm gamma/beta. Best practice is to exclude BatchNorm (and bias) parameters from weight decay, as regularizing them hurts convergence.

## Current State

```java
// AdamW.step() — applies same weightDecay to every parameter
for (var layer : layers) {
    var params = layer.parameters();
    var grads = layer.gradients();
    for (int i = 0; i < params.size(); i++) {
        computeOps.adamWUpdate(params.get(i), grads.get(i), m, v,
            lr, beta1, beta2, epsilon,
            weightDecay,  // ← same for ALL params, including BN gamma/beta
            bc1, bc2);
    }
}
```

## Why This Matters

Weight decay penalizes large parameter values. For BN's gamma (scale) and beta (shift), this is counterproductive:
- **gamma** should be free to scale features as needed (often >1.0)
- **beta** should be free to shift features
- Applying weight decay pushes gamma toward 0, effectively disabling BatchNorm

All major frameworks use parameter groups to exempt BN and bias from weight decay.

### Parameter Group Concept

```java
public record ParameterGroup(
    List<Tensor> parameters,
    List<Tensor> gradients,
    float weightDecay           // group-specific weight decay override
) {}

// Optimizer interface extension
public interface Optimizer {
    void step(List<ParametricLayer> layers);
    
    // NEW: step with parameter groups for fine-grained control
    default void step(List<ParameterGroup> groups) {
        // Default: delegate to step(layers) for backward compatibility
    }
}
```

### Automatic BN/Bias Detection

```java
// In Trainer or NetworkBuilder: auto-split parameters
public static List<ParameterGroup> autoGroupParameters(
        List<ParametricLayer> layers, float defaultWeightDecay) {
    
    List<Tensor> decayParams = new ArrayList<>();
    List<Tensor> decayGrads = new ArrayList<>();
    List<Tensor> noDecayParams = new ArrayList<>();
    List<Tensor> noDecayGrads = new ArrayList<>();
    
    for (var layer : layers) {
        if (layer instanceof BatchNorm || layer instanceof LayerNorm) {
            noDecayParams.addAll(layer.parameters());  // gamma, beta
            noDecayGrads.addAll(layer.gradients());
        } else {
            // For Dense/Conv2d: weights get decay, biases don't
            var params = layer.parameters();
            var grads = layer.gradients();
            decayParams.add(params.get(0));     // weights
            decayGrads.add(grads.get(0));
            if (params.size() > 1) {
                noDecayParams.add(params.get(1));  // bias
                noDecayGrads.add(grads.get(1));
            }
        }
    }
    
    return List.of(
        new ParameterGroup(decayParams, decayGrads, defaultWeightDecay),
        new ParameterGroup(noDecayParams, noDecayGrads, 0f)  // no decay
    );
}
```

### Usage

```java
var optimizer = new AdamW(0.001f, 0.9f, 0.999f, 1e-8f, 0.01f);

var trainer = new Trainer(model, loss, optimizer)
    .withParameterGrouping(true);  // auto-split BN/bias from weight decay
```

## Acceptance Criteria

- [ ] `ParameterGroup` record with group-specific weight decay
- [ ] `AdamW.step(List<ParameterGroup>)` respects per-group weight decay
- [ ] Auto-detection of BN/LayerNorm parameters for zero weight decay
- [ ] Auto-detection of bias parameters for zero weight decay
- [ ] Default behavior (no groups) unchanged
- [ ] SGD with weight decay also supports parameter groups
- [ ] Unit test: BN parameters receive zero weight decay
- [ ] Unit test: Conv2d weights receive configured weight decay
- [ ] Unit test: Conv2d biases receive zero weight decay

## References

- [Decoupled Weight Decay (Loshchilov & Hutter, 2019)](https://arxiv.org/abs/1711.05101) — Recommends excluding BN from decay
- [PyTorch parameter groups](https://pytorch.org/docs/stable/optim.html#per-parameter-options)
- [ResNet training recipes](https://github.com/pytorch/vision/blob/main/references/classification/train.py) — Always groups parameters
